local_storage.go

```diff
@@ -33,66 +33,141 @@ func newItemNotFoundError(format string, args ...interface{}) ItemNotFoundError
 	}
 }
 
-// SetItem saves the given item in localStorage under the given key.
-func SetItem(key, item string) error {
+func init() {
+	DetectStorage()
+}
+
+// DetectStorage detects and (re)initializes the localStorage.
+func DetectStorage() (ok bool) {
+	defer func() {
+		if r := recover(); r != nil {
+			localStorage = nil
+			ok = false
+		}
+	}()
+
+	localStorage = js.Global.Get("localStorage")
+
 	if localStorage == js.Undefined {
+		localStorage = nil
+	}
+	if localStorage == nil {
+		return false
+	}
+
+	// Cf. https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API
+	// https://gist.github.com/paulirish/5558557
+	x := "__storage_test__"
+	localStorage.Set(x, x)
+	obj := localStorage.Get(x)
+	if obj == js.Undefined || obj == nil {
+		localStorage = nil
+		return false
+	}
+	localStorage.Call("removeItem", x)
+	return true
+}
+
+// SetItem saves the given item in localStorage under the given key.
+func SetItem(key, item string) (err error) {
+	if localStorage == nil || localStorage == js.Undefined {
 		return ErrLocalStorageNotSupported
 	}
+	defer func() {
+		if r := recover(); r != nil {
+			err = r.(error)
+		}
+	}()
+
 	localStorage.Call("setItem", key, item)
 	return nil
 }
 
 // GetItem finds and returns the item identified by key. If there is no item in
 // localStorage with the given key, GetItem will return an ItemNotFoundError.
-func GetItem(key string) (string, error) {
-	if localStorage == js.Undefined {
+func GetItem(key string) (s string, err error) {
+	if localStorage == nil || localStorage == js.Undefined {
 		return "", ErrLocalStorageNotSupported
 	}
+	defer func() {
+		if r := recover(); r != nil {
+			err = r.(error)
```

Comment:
You're right about the recover type, I'll add a type assertion.

Regarding your 2nd point, I can add an explicit s=0 but AFAICS the patch already implements the behavior you're expecting, since s is only set if the js call is successful.